### PR TITLE
Ease the creation of others sub-commands for cli

### PR DESCRIPTION
**Pull request by rlespinasse**
Internally the cli configuration is done with the Service Provider API.

Now, all classes who implements `CliCommand` and are declared as implementation for the service provider API are added to the cli automatically.

All current sub-commands implements `CliCommand`.

### CODE REVIEW

File: src/main/java/fr/insalyon/citi/golo/cli/Main.java [hunk 1]
```diff
@@ -201,30 +73,13 @@ public static void main(String... args) throws Throwable {
...
-          case "new":
-            init(init);
-            break;
-          default:
-            throw new AssertionError("WTF?");
+        String parsedCommand = cmd.getParsedCommand();
+        JCommander parsedJCommander = cmd.getCommands().get(parsedCommand);
+        Object commandObject = parsedJCommander.getObjects().get(0);
+        if(commandObject instanceof CliCommand) {
+          ((CliCommand) commandObject).execute();
```
Comment (jponge) [inline src/main/java/fr/insalyon/citi/golo/cli/Main.java]: Isn't there a variant with generics?

Comment (rlespinasse) [inline src/main/java/fr/insalyon/citi/golo/cli/Main.java]: JCommander object love `Object` as generic type :smile: 
https://github.com/cbeust/jcommander/blob/master/src/main/java/com/beust/jcommander/JCommander.java#L1451-L1453

Comment (jponge) [inline src/main/java/fr/insalyon/citi/golo/cli/Main.java]: That's type unsafe :-)

Comment (rlespinasse) [inline src/main/java/fr/insalyon/citi/golo/cli/Main.java]: With that, no need of
```java
@SuppressWarnings("unchecked")
```
:smiley:

**Comment by jponge** [comment]
Looks good to me. Ping @danielpetisme

**Comment by yloiseau** [comment]
no comment... but I see conflicts comming :sweat:

**Comment by rlespinasse** [comment]
This PR can wait, and I can deal with the conflicts :smile:, can you tell me which PR can be in conflict?

**Comment by yloiseau** [comment]
Not yet in PR but in my macro wip branch, but I don't think it will be. too hard to resolve

**Comment by rlespinasse** [comment]
I check your code on https://github.com/yloiseau/golo-lang/blob/wip/macro-quote, and I am of the same opinion as you on the difficulty of resolution

**Comment by jponge** [comment]
Don't be too pessimistic guys :-)

**Comment by yloiseau** [comment]
And indeed your refactoring is realy nice

**Comment by yloiseau** [comment]
Moreover my PR is not ready yet

**Comment by jponge** [comment]
All good to me.

**Comment by jponge** [comment]
Just tried on a local copy and I have a bug :-)

```
$ target/golo-3.0.0-SNAPSHOT-distribution/golo-3.0.0-SNAPSHOT/bin/golo diagnose --tool ast oppriority.golo
Can't instantiate validator:java.lang.InstantiationException: fr.insalyon.citi.golo.cli.command.DiagnoseCommand$DiagnoseModeValidator

Diagnosis for the Golo compiler internals
Usage: diagnose [options] Golo source files (*.golo and directories)
  Options:
    --tool
       The diagnosis tool to use: {ast, ir}
       Default: ir
```

**Comment by rlespinasse** [comment]
it's link to a package protected class (a.k.a hidden for the `Main` class point of view)
https://github.com/golo-lang/golo-lang/blob/master/src/main/java/fr/insalyon/citi/golo/cli/command/DiagnoseCommand.java#L124

```shell
$ golo diagnose
(OK)
$ golo diagnose --tool ast ...
Can't instantiate validator:java.lang.InstantiationException: fr.insalyon.citi.golo.cli.command.DiagnoseCommand$DiagnoseModeValidator
```

Which do you prefer?
- `public static` inner classes
- `public` classes

Two commands in error : `golo diagnose` and `golo doc`

**Comment by jponge** [comment]
`public` static inner classes sound good to me.

**Pull request merged**